File: ios/PersonalSite/Features/OpenSource/RepositoryBrowserView.swift
import SwiftUI

/// 仓库目录页：经 /api/open-source/[slug]/repository/tree 拉全量树（服务端有缓存），
/// 每层按路径前缀过滤出直接子项，目录在前。
struct RepositoryDirectoryView: View {
    let slug: String
    let path: String

    @State private var entries: [RepositoryTreeEntry] = []
    @State private var isLoading = true
    @State private var errorMessage: String?
    @State private var truncated = false

    var body: some View {
        LoadStateView(
            isLoading: isLoading,
            errorMessage: errorMessage,
            isEmpty: children.isEmpty,
            emptyMessage: "目录为空",
            onRetry: { Task { await load() } }
        ) {
            List {
                if truncated {
                    Text("仓库较大，仅展示部分条目")
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }
                ForEach(children, id: \.path) { entry in
                    if entry.type == .tree {
                        NavigationLink(value: OpenSourceRoute.directory(slug: slug, path: entry.path)) {
                            Label(entry.name, systemImage: "folder")
                        }
                    } else {
                        NavigationLink(value: OpenSourceRoute.file(slug: slug, path: entry.path)) {
                            HStack {
                                Label(entry.name, systemImage: "doc.text")
                                Spacer()
                                if let size = entry.size {
                                    Text(Self.sizeText(size))
                                        .font(.caption)
                                        .foregroundStyle(.secondary)
                                }
                            }
                        }
                    }
                }
            }
            .listStyle(.plain)
        }
        .navigationTitle(path.isEmpty ? "仓库" : path.split(separator: "/").last.map(String.init) ?? path)
        .navigationBarTitleDisplayMode(.inline)
        .task { await load() }
    }

    /// 直接子项：路径以当前目录为前缀且剩余部分不含 "/"。
    private var children: [RepositoryTreeEntry] {
        let prefix = path.isEmpty ? "" : path + "/"
        return entries
            .filter { entry in
                guard entry.path.hasPrefix(prefix) else { return false }
                return !entry.path.dropFirst(prefix.count).contains("/")
            }
            .sorted { left, right in
                if (left.type == .tree) != (right.type == .tree) { return left.type == .tree }
                return left.name.localizedStandardCompare(right.name) == .orderedAscending
            }
    }

    private func load() async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            let tree: RepositoryTreeResponse = try await SiteAPIClient().get("/api/open-source/\(slug)/repository/tree")
            entries = tree.entries
            truncated = tree.truncated
        } catch {
            errorMessage = "读取仓库目录失败，请稍后重试。"
        }
    }

    private static func sizeText(_ bytes: Int) -> String {
        if bytes < 1024 { return "\(bytes) B" }
        return String(format: "%.1f KB", Double(bytes) / 1024.0)
    }
}

private extension RepositoryTreeEntry {
    var name: String {
        path.split(separator: "/").last.map(String.init) ?? path
    }
}

/// 仓库文件页：文本内容等宽展示；二进制给 GitHub 原文件链接。
struct RepositoryFileView: View {
    let slug: String
    let path: String

    @State private var file: RepositoryFileResponse?
    @State private var errorMessage: String?

    private var stateIdentity: LoadStateIdentity {
        file != nil ? .content : errorMessage != nil ? .error : .loading
    }

    var body: some View {
        ZStack {
            Group {
            if let file {
                if file.binary {
                    ContentUnavailableView {
                        Label("二进制文件，无法预览", systemImage: "doc")
                    } actions: {
                        if let url = URL(string: file.fileUrl) {
                            Link("在 GitHub 查看", destination: url)
                        }
                    }
                } else {
                    ScrollView([.vertical, .horizontal]) {
                        Text(file.content ?? "")
                            .font(.system(.footnote, design: .monospaced))
                            .textSelection(.enabled)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding()
                    }
                }
            } else if let errorMessage {
                ContentUnavailableView {
                    Label("加载失败", systemImage: "exclamationmark.triangle")
                } description: {
                    Text(errorMessage)
                } actions: {
                    Button("重试") { Task { await load() } }
                }
            } else {
                ProgressView()
            }
            }
            .id(stateIdentity)
            .transition(.opacity)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .animation(PSMotion.stateChange, value: stateIdentity)
        .navigationTitle(path.split(separator: "/").last.map(String.init) ?? path)
        .navigationBarTitleDisplayMode(.inline)
        .task { await load() }
    }

    private func load() async {
        errorMessage = nil
        do {
            file = try await SiteAPIClient().get(
                "/api/open-source/\(slug)/repository/file",
                queryItems: [URLQueryItem(name: "path", value: path)]
            )
        } catch {
            errorMessage = "读取文件失败，请稍后重试。"
        }
    }
}

File: android/app/src/main/java/com/personalsite/features/opensource/RepositoryBrowserView.kt
package com.personalsite.features.opensource

import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Description
import androidx.compose.material.icons.filled.Folder
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.personalsite.core.PSColors
import com.personalsite.core.SiteApiClient
import com.personalsite.features.support.LoadStateView
import com.personalsite.models.RepositoryFileResponse
import com.personalsite.models.RepositoryTreeEntry
import com.personalsite.models.RepositoryTreeResponse
import java.util.Locale
import kotlinx.coroutines.launch

/** 页面小标题（对齐 iOS navigationTitle）：目录 path 为空显示「仓库」，否则最后一段；文件取最后一段。 */
@Composable
private fun BrowserHeader(title: String) {
    Text(
        text = title,
        fontSize = 17.sp,
        fontWeight = FontWeight.SemiBold,
        color = PSColors.ink,
        modifier = Modifier.padding(horizontal = 16.dp, vertical = 12.dp),
    )
}

/**
 * 仓库目录页：经 /api/open-source/{slug}/repository/tree 拉全量树（服务端有缓存），
 * 每层按路径前缀过滤出直接子项，目录在前。
 */
@Composable
fun RepositoryDirectoryView(
    slug: String,
    path: String,
    onOpenDirectory: (String, String) -> Unit,
    onOpenFile: (String, String) -> Unit,
) {
    var entries by remember { mutableStateOf<List<RepositoryTreeEntry>>(emptyList()) }
    var truncated by remember { mutableStateOf(false) }
    var isLoading by remember { mutableStateOf(true) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()
    val api = remember { SiteApiClient() }

    suspend fun load() {
        isLoading = true
        errorMessage = null
        try {
            val tree: RepositoryTreeResponse = api.get("/api/open-source/$slug/repository/tree")
            entries = tree.entries
            truncated = tree.truncated
        } catch (e: Exception) {
            errorMessage = "读取仓库目录失败，请稍后重试。"
        }
        isLoading = false
    }

    LaunchedEffect(slug) { load() }

    // 直接子项：路径以当前目录为前缀且剩余部分不含 "/"
    val prefix = if (path.isEmpty()) "" else "$path/"
    val children = entries
        .filter { it.path.startsWith(prefix) && !it.path.substring(prefix.length).contains('/') }
        .sortedWith(compareBy({ it.type != RepositoryTreeEntry.EntryType.TREE }, { it.name }))

    Column(Modifier.fillMaxSize()) {
        BrowserHeader(title = if (path.isEmpty()) "仓库" else path.substringAfterLast('/'))
        LoadStateView(
            isLoading = isLoading,
            errorMessage = errorMessage,
            isEmpty = children.isEmpty(),
            emptyMessage = "目录为空",
            onRetry = { scope.launch { load() } },
        ) {
            LazyColumn(Modifier.fillMaxSize()) {
                if (truncated) {
                    item(key = "truncated-hint") {
                        Text(
                            text = "仓库较大，仅展示部分条目",
                            fontSize = 12.sp,
                            color = PSColors.quiet,
                            modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp),
                        )
                    }
                }
                items(children, key = { it.path }) { entry ->
                    if (entry.type == RepositoryTreeEntry.EntryType.TREE) {
                        DirectoryRow(name = entry.name, onClick = { onOpenDirectory(slug, entry.path) })
                    } else {
                        FileRow(
                            name = entry.name,
                            sizeText = entry.size?.let(::sizeText),
                            onClick = { onOpenFile(slug, entry.path) },
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun DirectoryRow(name: String, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            imageVector = Icons.Filled.Folder,
            contentDescription = null,
            tint = PSColors.ink,
            modifier = Modifier.size(18.dp),
        )
        Spacer(Modifier.width(8.dp))
        Text(name, color = PSColors.ink)
    }
}

@Composable
private fun FileRow(name: String, sizeText: String?, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            imageVector = Icons.Filled.Description,
            contentDescription = null,
            tint = PSColors.ink,
            modifier = Modifier.size(18.dp),
        )
        Spacer(Modifier.width(8.dp))
        Text(name, color = PSColors.ink, modifier = Modifier.weight(1f))
        if (sizeText != null) {
            Text(sizeText, fontSize = 12.sp, color = PSColors.quiet)
        }
    }
}

private fun sizeText(bytes: Long): String =
    if (bytes < 1024) "$bytes B" else String.format(Locale.US, "%.1f KB", bytes / 1024.0)

/** 仓库文件页：文本内容等宽展示；二进制给 GitHub 原文件链接。 */
@Composable
fun RepositoryFileView(slug: String, path: String) {
    var file by remember { mutableStateOf<RepositoryFileResponse?>(null) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()
    val api = remember { SiteApiClient() }
    val uriHandler = LocalUriHandler.current

    suspend fun load() {
        errorMessage = null
        try {
            file = api.get(
                "/api/open-source/$slug/repository/file",
                query = mapOf("path" to path),
            )
        } catch (e: Exception) {
            errorMessage = "读取文件失败，请稍后重试。"
        }
    }

    LaunchedEffect(slug, path) { load() }

    Column(Modifier.fillMaxSize()) {
        BrowserHeader(title = path.substringAfterLast('/'))
        LoadStateView(
            isLoading = file == null && errorMessage == null,
            errorMessage = errorMessage,
            isEmpty = false,
            emptyMessage = "",
            onRetry = { scope.launch { load() } },
        ) {
            file?.let { current ->
                if (current.binary) {
                    Box(Modifier.fillMaxSize().padding(24.dp)) {
                        Column(
                            modifier = Modifier.align(Alignment.Center),
                            horizontalAlignment = Alignment.CenterHorizontally,
                        ) {
                            Icon(
                                imageVector = Icons.Filled.Description,
                                contentDescription = null,
                                tint = PSColors.quiet,
                            )
                            Spacer(Modifier.height(8.dp))
                            Text("二进制文件，无法预览", color = PSColors.ink, fontWeight = FontWeight.SemiBold)
                            Spacer(Modifier.height(12.dp))
                            Button(onClick = { uriHandler.openUri(current.fileUrl) }) {
                                Text("在 GitHub 查看")
                            }
                        }
                    }
                } else {
                    SelectionContainer(
                        Modifier
                            .fillMaxSize()
                            .verticalScroll(rememberScrollState())
                            .horizontalScroll(rememberScrollState()),
                    ) {
                        Text(
                            text = current.content.orEmpty(),
                            fontFamily = FontFamily.Monospace,
                            fontSize = 13.sp,
                            color = PSColors.ink,
                            modifier = Modifier.padding(16.dp),
                        )
                    }
                }
            }
        }
    }
}
